File: problems/2022_09_27/0039/Solution.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> list = new ArrayList<>();
        backtrack(0, target, candidates, res, list);
        return res;
    }

    private void backtrack(int i, int target, int[] candidates, List<List<Integer>> res, List<Integer> list) {
        if (target == 0) {
            res.add(new ArrayList<>(list));
        } else if (target < 0 || i >= candidates.length) {
            return;
        } else {
            list.add(candidates[i]);
            backtrack(i, target - candidates[i], candidates, res, list);

            list.remove(list.size() - 1);
            backtrack(i + 1, target, candidates, res, list);
        }
    }
}

File: problems/2022_09_27/0039/Solution.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        self.res = []
        self.backtracking(candidates, target, [], 0, 0)
        return self.res

    def backtracking(self, candidates, target, curr, total, index):
        if total == target:
            self.res.append(curr.copy())
            return

        if index >= len(candidates) or total > target:
            return

        curr.append(candidates[index])
        self.backtracking(candidates, target, curr,
                          total + candidates[index], index)

        curr.pop()
        self.backtracking(candidates, target, curr, total, index + 1)
